blocks/api/parser.js

```diff
@@ -181,6 +181,28 @@ export function createBlockWithFallback( name, innerHTML, attributes ) {
 		// as invalid, or future serialization attempt results in an error
 		block.originalContent = innerHTML;
 
+		// When a block is invalid, attempt to validate again using a supplied `deprecated` definition.
+		// This allows blocks to modify their attribute and markup structure without invalidating
+		// content written in previous formats.
+		if ( ! block.isValid && blockType.deprecated ) {
+			let attributesParsedWithDeprecatedVersion;
+			const hasValidOlderVersion = find( blockType.deprecated, ( oldBlockType ) => {
+				const deprecatedBlockType = {
+					...omit( blockType, [ 'attributes', 'save', 'supports' ] ), // Parsing/Serialization properties
+					...oldBlockType,
+				};
+				const deprecatedBlockAttributes = getBlockAttributes( deprecatedBlockType, innerHTML, attributes );
+				const isValid = isValidBlock( innerHTML, deprecatedBlockType, deprecatedBlockAttributes );
+				attributesParsedWithDeprecatedVersion = isValid ? deprecatedBlockAttributes : undefined;
```

Comment:
If we did extract this to a separate method, I could see this working well as an early return from a loop:

```js
for ( let i = 0; i < deprecated.length; i++ ) {
	// ...
	if ( isValid ) {
		return deprecatedBlockAttributes;
	}
}
```